File: src/common/structs/BitArray.cpp
/*
 * euler - A collection of ProjectEuler libraries, tools, and solutions.
 * Copyright (C) 2013 Axel Rasmussen
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "BitArray.hpp"

#include <cassert>
#include <iomanip>
#include <sstream>
#include <stdexcept>
#include <utility>

#include <bdrck/algorithm/String.hpp>

#include "common/math/Math.hpp"
#include "common/util/Bitwise.hpp"

namespace
{
void arrayDeleter(uint64_t *a)
{
	if(a != nullptr)
		delete[] a;
}

std::pair<std::size_t, std::size_t> bitToChunkIndex(std::size_t p)
{
	return euler::math::divide(p, euler::structs::BitArray::CHUNK_BIT);
}

std::size_t getMinimumChunkSize(std::size_t s)
{
	if(s == 0)
		return 0;
	return bitToChunkIndex(s - 1).first + 1;
}

bool getRaw(euler::structs::BitArray::chunk_type const *data, std::size_t p)
{
	auto index = bitToChunkIndex(p);
	auto mask = euler::structs::BitArray::chunk_type(1) << index.second;
	return data[index.first] & mask;
}

void setRaw(euler::structs::BitArray::chunk_type *data, std::size_t p, bool v)
{
	auto index = bitToChunkIndex(p);
	auto mask = euler::structs::BitArray::chunk_type(1) << index.second;
	if(v)
		data[index.first] |= mask;
	else
		data[index.first] &= ~mask;
}

/*!
 * Shift the chunks in the given chunk array left by n places. Fill in n
 * chunks with 0's at the bottom of the chunk array.
 *
 * \param data The chunk array to shift.
 * \param l The length of the chunk array.
 * \param n The number of places (chunks) to shift left.
 */
void shiftChunksLeft(euler::structs::BitArray::chunk_type *data, std::size_t l,
                     std::size_t n)
{
	n = std::min(n, l);
	// Compute the number of chunks that will be shifted off the left end.
	std::size_t shiftedOff = n + n - std::min(l, n + n);
	// Setup the pointers to where the chunk copies will occur.
	auto const *srcFirst = data;
	auto const *srcLast = data + n - shiftedOff;
	auto *dstLast = data + std::min(l, n + n);
	// Copy chunks starting from the end of the chunk array.
	std::copy_backward(srcFirst, srcLast, dstLast);
	// Fill in the bottom n chunks with 0's.
	std::fill(data, data + n, 0);
}

/*!
 * Shift the chunks in the given chunk array right by n places. Fill in n
 * chunks with 0's at the top of the chunk array.
 *
 * \param data The chunk array to shift.
 * \param l The length of the chunk array.
 * \param n The number of places (chunks) to shift right.
 */
void shiftChunksRight(euler::structs::BitArray::chunk_type *data, std::size_t l,
                      std::size_t n)
{
	n = std::min(n, l);
	// Compute the number of chunks that will be shifted off the right end.
	std::size_t shiftedOff = n + n - std::min(l, n + n);
	// Setup the pointers to where the chunk copies will occur.
	auto const *srcFirst = data + l - n + shiftedOff;
	auto const *srcLast = data + l;
	auto *dstFirst = data;
	// Copy chunks starting from the front of the chunk array.
	std::copy(srcFirst, srcLast, dstFirst);
	// Fill in the top n chunks with 0's.
	std::fill(data + l - n, data + l, 0);
}
}

namespace euler
{
namespace structs
{
BitArray::BitArray(std::size_t s, BitArrayFillPolicy p)
        : bitSize(0), chunkSize(0), data(nullptr, arrayDeleter)
{
	resize(s, p);
}

BitArray::BitArray(BitArray const &o)
        : bitSize(0), chunkSize(0), data(nullptr, arrayDeleter)
{
	*this = o;
}

BitArray &BitArray::operator=(BitArray const &o)
{
	if(this == &o)
		return *this;
	bitSize = o.bitSize;
	chunkSize = o.chunkSize;
	data.reset(new chunk_type[chunkSize]);
	std::copy(o.data.get(), o.data.get() + o.chunkSize, data.get());
	return *this;
}

int BitArray::compare(BitArray const &o) const
{
	std::size_t maxSize = std::max(size(), o.size());
	for(std::size_t offFromEnd = 0; offFromEnd < maxSize; ++offFromEnd)
	{
		std::size_t p = maxSize - 1 - offFromEnd;
		if(test(p, false) && !o.test(p, false))
			return 1;
		else if(!test(p, false) && o.test(p, false))
			return -1;
	}
	return 0;
}

bool BitArray::operator==(BitArray const &o) const
{
	return compare(o) == 0;
}

bool BitArray::operator!=(BitArray const &o) const
{
	return !(*this == o);
}

bool BitArray::operator<(BitArray const &o) const
{
	return compare(o) < 0;
}

bool BitArray::operator<=(BitArray const &o) const
{
	return compare(o) <= 0;
}

bool BitArray::operator>(BitArray const &o) const
{
	return compare(o) > 0;
}

bool BitArray::operator>=(BitArray const &o) const
{
	return compare(o) >= 0;
}

std::size_t BitArray::size() const
{
	return bitSize;
}

std::size_t BitArray::byteSize() const
{
	assert(bitSize % 8 == 0);
	return bitSize / 8;
}

void BitArray::resize(std::size_t s, BitArrayFillPolicy p)
{
	if(s == 0)
	{
		bitSize = 0;
		chunkSize = 0;
		data.reset();
		return;
	}

	auto newChunkSize = getMinimumChunkSize(s);
	decltype(data) newData(new chunk_type[newChunkSize], arrayDeleter);

	chunk_type fill(0);
	if(p == BitArrayFillPolicy::SET ||
	   p == BitArrayFillPolicy::PRESERVE_SET)
	{
		fill = ~fill;
	}

	for(std::size_t i = 0; i < newChunkSize; ++i)
	{
		chunk_type v = fill;
		if(p == BitArrayFillPolicy::PRESERVE_SET ||
		   p == BitArrayFillPolicy::PRESERVE_CLEAR)
		{
			auto chunk = chunkAt(i);
			v = (chunk.first & chunk.second) |
			    (fill & ~chunk.second);
		}
		newData.get()[i] = v;
	}

	bitSize = s;
	chunkSize = newChunkSize;
	data = std::move(newData);
}

std::size_t BitArray::capacity() const
{
	return chunkSize * CHUNK_BIT;
}

std::size_t BitArray::byteCapacity() const
{
	return capacity() / 8;
}

std::pair<BitArray::chunk_type, BitArray::chunk_type>
BitArray::chunkAt(std::size_t p) const
{
	if(p >= chunkSize)
		return std::make_pair<chunk_type, chunk_type>(0, 0);

	chunk_type chunk = data.get()[p];
	chunk_type mask = ~chunk_type(0);

	// If this is the last chunk, compute a mask.
	if(p == chunkSize - 1)
	{
		std::size_t lastChunkBits = bitSize % CHUNK_BIT;
		if(lastChunkBits == 0)
			lastChunkBits = CHUNK_BIT;
		mask = mask << (CHUNK_BIT - lastChunkBits) >>
		       (CHUNK_BIT - lastChunkBits);
	}

	return std::make_pair(chunk, mask);
}

bool BitArray::operator[](std::size_t p) const
{
	return getRaw(data.get(), p);
}

bool BitArray::test(std::size_t p, boost::optional<bool> defaultValue) const
{
	if(p >= bitSize)
	{
		if(!!defaultValue)
		{
			return *defaultValue;
		}
		else
		{
			throw std::runtime_error(
			        "BitArray index is out of bounds.");
		}
	}
	return (*this)[p];
}

bool BitArray::at(std::size_t p, boost::optional<bool> defaultValue) const
{
	return test(p, defaultValue);
}

bool BitArray::all() const
{
	constexpr chunk_type ALL_CHUNK = ~chunk_type(0);
	for(std::size_t i = 0; i < chunkSize; ++i)
	{
		auto chunk = chunkAt(i);
		if((chunk.first & chunk.second) != (ALL_CHUNK & chunk.second))
			return false;
	}
	return true;
}

bool BitArray::any() const
{
	return !none();
}

bool BitArray::none() const
{
	for(std::size_t i = 0; i < chunkSize; ++i)
	{
		auto chunk = chunkAt(i);
		if((chunk.first & chunk.second) != chunk_type(0))
			return false;
	}
	return true;
}

std::size_t BitArray::count() const
{
	std::size_t c = 0;
	for(std::size_t i = 0; i < chunkSize; ++i)
		c += euler::util::bitwise::opop(data.get()[i]);
	return c;
}

BitArray &BitArray::set()
{
	for(std::size_t i = 0; i < chunkSize; ++i)
		data.get()[i] = ~chunk_type(0);
	return *this;
}

BitArray &BitArray::set(std::size_t p, bool v)
{
	setRaw(data.get(), p, v);
	return *this;
}

BitArray &BitArray::reset()
{
	for(std::size_t i = 0; i < chunkSize; ++i)
		data.get()[i] = chunk_type(0);
	return *this;
}

BitArray &BitArray::reset(std::size_t p)
{
	return set(p, false);
}

BitArray &BitArray::flip()
{
	std::transform(data.get(), data.get() + chunkSize, data.get(),
	               [](chunk_type c) -> chunk_type { return ~c; });
	return *this;
}

BitArray &BitArray::flip(std::size_t p)
{
	auto index = bitToChunkIndex(p);
	chunk_type mask = chunk_type(1) << index.second;
	if(data.get()[index.first] & mask)
		data.get()[index.first] &= ~mask;
	else
		data.get()[index.first] |= mask;
	return *this;
}

BitArray BitArray::operator<<(std::size_t n) const
{
	BitArray a(*this);
	a <<= n;
	return a;
}

BitArray &BitArray::operator<<=(std::size_t n)
{
	// If this array is empty, just return.
	if(size() == 0)
		return *this;

	// If we're shifting the whole thing, just clear.
	if(n >= size())
	{
		reset();
		return *this;
	}

	// Shift as many full chunks as possible first.
	shiftChunksLeft(data.get(), chunkSize, n / CHUNK_BIT);

	// Shift the remaining number of bits left in each chunk.
	std::size_t bitShift = n % CHUNK_BIT;
	for(std::size_t idx = chunkSize - 1; idx > 0; --idx)
	{
		data.get()[idx] <<= bitShift;
		data.get()[idx] |=
		        data.get()[idx - 1] >> (CHUNK_BIT - bitShift);
	}
	data.get()[0] <<= bitShift;

	return *this;
}

BitArray BitArray::operator>>(std::size_t n) const
{
	BitArray a(*this);
	a >>= n;
	return a;
}

BitArray &BitArray::operator>>=(std::size_t n)
{
	// If this array is empty, just return.
	if(size() == 0)
		return *this;

	// If we're shifting the whole thing, just clear.
	if(n >= size())
	{
		reset();
		return *this;
	}

	// Shift as many full chunks as possible first.
	shiftChunksRight(data.get(), chunkSize, n / CHUNK_BIT);

	// Shift the remaining number of bits right in each chunk.
	std::size_t bitShift = n % CHUNK_BIT;
	for(std::size_t idx = 0; idx < (chunkSize - 1); ++idx)
	{
		data.get()[idx] >>= bitShift;
		data.get()[idx] |= data.get()[idx + 1]
		                   << (CHUNK_BIT - bitShift);
	}
	data.get()[chunkSize - 1] >>= bitShift;

	return *this;
}

BitArray &BitArray::leftRotate(std::size_t n)
{
	// v <<< n = (v << n) | (v >> (bits - n))

	BitArray b(*this);
	// Remove any whole multiples of our size (full rotations) from n.
	n %= size();
	*this <<= n;
	b >>= size() - n;
	*this |= b;
	return *this;
}

BitArray &BitArray::rightRotate(std::size_t n)
{
	// v >>> n = (v >> n) | (v << (bits - n))

	BitArray b(*this);
	// Remove any whole multiples of our size (full rotations) from n.
	n %= size();
	*this >>= n;
	b <<= size() - n;
	*this |= b;
	return *this;
}

BitArray BitArray::operator&(BitArray const &o) const
{
	BitArray a(*this);
	a &= o;
	return a;
}

BitArray &BitArray::operator&=(BitArray const &o)
{
	for(std::size_t i = 0; i < chunkSize; ++i)
	{
		data.get()[i] &=
		        i < o.chunkSize ? o.data.get()[i] : chunk_type(0);
	}
	return *this;
}

BitArray BitArray::operator|(BitArray const &o) const
{
	BitArray a(*this);
	a |= o;
	return a;
}

BitArray &BitArray::operator|=(BitArray const &o)
{
	for(std::size_t i = 0; i < chunkSize; ++i)
	{
		data.get()[i] |=
		        i < o.chunkSize ? o.data.get()[i] : chunk_type(0);
	}
	return *this;
}

BitArray BitArray::operator^(BitArray const &o) const
{
	BitArray a(*this);
	a ^= o;
	return a;
}

BitArray &BitArray::operator^=(BitArray const &o)
{
	for(std::size_t i = 0; i < chunkSize; ++i)
	{
		data.get()[i] ^=
		        i < o.chunkSize ? o.data.get()[i] : chunk_type(0);
	}
	return *this;
}

BitArray BitArray::operator~() const
{
	return BitArray(*this).flip();
}

std::string BitArray::to_string() const
{
	std::ostringstream oss;
	for(std::size_t idx = 0; idx < size(); ++idx)
		oss << ((*this)[idx] ? '1' : '0');
	return oss.str();
}

unsigned long BitArray::to_ulong() const
{
	return static_cast<unsigned long>(to_uintmax());
}

unsigned long long BitArray::to_ullong() const
{
	return static_cast<unsigned long long>(to_uintmax());
}

uintmax_t BitArray::to_uintmax() const
{
	uintmax_t value(0);

	// Build a vector of chunk values which will make up the uintmax_t.
	constexpr std::size_t floorChunks =
	        sizeof(uintmax_t) / sizeof(chunk_type);
	constexpr std::size_t byteRemainder =
	        sizeof(uintmax_t) % sizeof(chunk_type);
	constexpr std::size_t totalChunkCount =
	        floorChunks + (byteRemainder > 0 ? 1 : 0);
	std::size_t chunkCount = std::min(totalChunkCount, chunkSize);
	std::vector<chunk_type> chunks(chunkCount, 0);
	for(std::size_t i = 0; i < chunkCount; ++i)
	{
		auto chunk = chunkAt(i);
		chunks[i] = chunk.first & chunk.second;
	}

	// Copy the chunk data from this vector to the uintmax_t.
	uint8_t const *srcFirst =
	        reinterpret_cast<uint8_t const *>(chunks.data());
	std::size_t byteLength =
	        std::min(sizeof(uintmax_t), chunkSize * sizeof(chunk_type));
	uint8_t const *srcLast = srcFirst + byteLength;
	uint8_t *dstFirst = reinterpret_cast<uint8_t *>(&value);
	std::copy(srcFirst, srcLast, dstFirst);

	return value;
}

std::ostream &operator<<(std::ostream &os, BitArray const &ba)
{
	os << ba.to_string();
	return os;
}
}
}

File: src/structs/bit_array.rs
// euler - A collection of ProjectEuler libraries, tools, and solutions.
// Copyright (C) 2013 Axel Rasmussen
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.


use math::division;
use std::iter::Iterator;
use std::vec::Vec;

pub enum FillPolicy {
    Clear,
    PreserveClear,
    PreserveSet,
    Set,
}

pub type Chunk = u64;

const CHUNK_BIT: usize = 64;
const ALL_CHUNK: Chunk = !(0 as Chunk);

/// Returns the index of the given bit as a (chunk index, bit offset) tuple.
fn bit_to_chunk_index(index: usize) -> (usize, usize) { division::divide(index, CHUNK_BIT) }

fn get_minimum_chunk_size(size: usize) -> usize {
    if size == 0 {
        return 0;
    }
    bit_to_chunk_index(size - 1).0 + 1
}

#[derive(Clone)]
pub struct BitArray {
    bits: usize,
    data: Vec<Chunk>,
}

impl BitArray {
    pub fn new(size: usize, policy: FillPolicy) -> BitArray {
        let mut arr = BitArray {
            bits: 0,
            data: Vec::new(),
        };
        arr.resize(size, policy);
        arr
    }

    pub fn len(&self) -> usize { self.bits }

    pub fn len_bytes(&self) -> usize { self.len() / 8 }

    pub fn len_chunks(&self) -> usize { self.data.len() }

    pub fn resize(&mut self, size: usize, policy: FillPolicy) {
        if size == 0 {
            self.bits = 0;
            self.data = Vec::new();
            return;
        }

        let new_chunk_size = get_minimum_chunk_size(size);
        let mut new_data: Vec<Chunk> = vec![0; new_chunk_size];

        let fill: Chunk = match policy {
            FillPolicy::Clear => 0,
            FillPolicy::PreserveClear => 0,
            _ => !(0 as Chunk),
        };

        for i in 0..new_chunk_size {
            *new_data.get_mut(i).unwrap() = match policy {
                FillPolicy::Clear => fill,
                FillPolicy::Set => fill,
                _ => {
                    let chunk = self.get_chunk(i).unwrap_or((0, 0));
                    (chunk.0 & chunk.1) | (fill & !chunk.1)
                },
            };
        }

        self.bits = size;
        self.data = new_data;
    }

    pub fn as_chunk_slice(&self) -> &[Chunk] { self.data.as_slice() }

    pub fn iter(&self) -> Iter { Iter::new(self) }

    /// Returns the chunk at the given position, along with a mask indicating
    /// which bits in the chunk are actually members of this bit array. If the
    /// given position is out of bounds, None is returned instead.
    pub fn get_chunk(&self, index: usize) -> Option<(Chunk, Chunk)> {
        let mut mask: Chunk = !(0 as Chunk);
        if index == self.data.len() - 1 {
            let rem_bits = self.bits % CHUNK_BIT;
            let last_chunk_bits = match rem_bits {
                0 => CHUNK_BIT,
                _ => rem_bits,
            };
            mask = mask << (CHUNK_BIT - last_chunk_bits) >> (CHUNK_BIT - last_chunk_bits);
        }

        self.data.get(index).map(|c| (*c, mask))
    }

    /// Access the bit at the given position, and return its value. If the
    /// given position is out-of-bounds, None is returned.
    pub fn get(&self, index: usize) -> Option<bool> {
        if index >= self.bits {
            return None;
        }
        let chunk_index = bit_to_chunk_index(index);
        Some(self.data.get(chunk_index.0).unwrap() & ((1 as Chunk) << chunk_index.1) != 0)
    }

    /// Sets the value of the bit at the given index to the given value. Panics
    /// if the given index is out of bounds.
    pub fn set(&mut self, index: usize, value: bool) {
        let chunk_index = bit_to_chunk_index(index);
        let mask: Chunk = (1 as Chunk) << chunk_index.1;
        let chunk: &mut Chunk = self.data.get_mut(chunk_index.0).unwrap();
        *chunk = match value {
            true => *chunk | mask,
            false => *chunk & !mask,
        }
    }

    /// Negate the value of the single bit at the given index. Like set(),
    /// panics if the given index is out of bounds.
    pub fn negate(&mut self, index: usize) {
        let value: bool = !self.get(index).unwrap();
        self.set(index, value);
    }

    pub fn set_all(&mut self) {
        for c in self.data.iter_mut() {
            *c = ALL_CHUNK;
        }
    }

    pub fn reset_all(&mut self) {
        for c in self.data.iter_mut() {
            *c = 0;
        }
    }

    pub fn negate_all(&mut self) {
        for c in self.data.iter_mut() {
            *c = !*c;
        }
    }

    /// Returns true if all bits are 1.
    pub fn all(&self) -> bool {
        for i in 0..self.data.len() {
            let chunk = self.get_chunk(i).unwrap();
            if (chunk.0 & chunk.1) != (ALL_CHUNK & chunk.1) {
                return false;
            }
        }
        true
    }

    /// Returns true if all bits are 0.
    pub fn none(&self) -> bool { !self.all() }
}

pub struct Iter<'a> {
    array: &'a BitArray,
    forward_index: usize,
    reverse_index: usize,
}

impl<'a> Iter<'a> {
    fn new(array: &'a BitArray) -> Iter<'a> {
        Iter {
            array: array,
            forward_index: 0,
            reverse_index: array.len(),
        }
    }
}

impl<'a> Iterator for Iter<'a> {
    type Item = (usize, bool);

    fn next(&mut self) -> Option<Self::Item> {
        let index = self.forward_index;
        let item = self.array.get(self.forward_index);
        self.forward_index += 1;
        item.map(|v| (index, v))
    }
}

impl<'a> DoubleEndedIterator for Iter<'a> {
    fn next_back(&mut self) -> Option<Self::Item> {
        if self.reverse_index == 0 {
            return None;
        }

        let index = self.reverse_index - 1;
        let item = self.array.get(self.reverse_index - 1);
        self.reverse_index -= 1;
        item.map(|v| (index, v))
    }
}
